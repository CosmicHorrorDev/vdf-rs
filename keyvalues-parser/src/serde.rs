use std::{borrow::Cow, collections::BTreeMap, fmt};

use crate::{Obj, Value};

use serde_core::{
    de::{MapAccess, Visitor},
    Deserialize, Deserializer,
};

fn value_string(s: impl ToString) -> Value<'static> {
    Value::Str(s.to_string().into())
}

struct ValueVisitor;

impl<'a> Visitor<'a> for ValueVisitor {
    type Value = Value<'a>;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("any valid VDF value")
    }

    fn visit_borrowed_str<E>(self, s: &'a str) -> Result<Self::Value, E> {
        Ok(Value::Str(Cow::Borrowed(s)))
    }

    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E> {
        Ok(value_string(s))
    }

    fn visit_string<E>(self, s: String) -> Result<Self::Value, E> {
        Ok(Value::Str(Cow::Owned(s)))
    }

    fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
    where
        V: MapAccess<'a>,
    {
        let mut obj = BTreeMap::new();
        while let Some((key, value)) = visitor.next_entry()? {
            obj.insert(key, value);
        }
        Ok(Value::Obj(crate::Obj(obj)))
    }
}

impl<'de: 'a, 'a> Deserialize<'de> for Value<'a> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(ValueVisitor)
    }
}

struct ObjVisitor;

impl<'a> Visitor<'a> for ObjVisitor {
    type Value = Obj<'a>;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("an object")
    }

    fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
    where
        V: MapAccess<'a>,
    {
        let mut obj = BTreeMap::new();
        while let Some((key, value)) = visitor.next_entry()? {
            obj.insert(key, value);
        }
        Ok(Obj(obj))
    }
}

impl<'de: 'a, 'a> Deserialize<'de> for Obj<'a> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_map(ObjVisitor)
    }
}
