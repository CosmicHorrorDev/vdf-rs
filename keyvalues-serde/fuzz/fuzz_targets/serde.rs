#![no_main]
use arbitrary::Arbitrary;
use keyvalues_serde::{from_str, to_string};
use libfuzzer_sys::fuzz_target;
use serde::{Deserialize, Serialize};

#[derive(Debug, PartialEq, Arbitrary, Deserialize, Serialize)]
struct KitchenSink {
    boolean: bool,
    character: char,
    float32: f32,
    float64: f64,
    signed08: i8,
    signed16: i16,
    signed32: i32,
    signed64: i64,
    unsigned08: u8,
    unsigned16: u16,
    unsigned32: u32,
    unsigned64: u64,
    // TODO: make a note about this
    #[serde(default)]
    vec: Vec<bool>,
    optional: Option<u32>,
    inner_struct: InnerStruct,
    inner_enum: InnerEnum,
    inner_tuple_struct: InnerTupleStruct,
}

#[derive(Debug, PartialEq, Arbitrary, Deserialize, Serialize)]
struct InnerStruct {
    field: String,
}

#[derive(Debug, PartialEq, Arbitrary, Deserialize, Serialize)]
enum InnerEnum {
    Foo,
    Bar,
    Baz,
}

#[derive(Debug, PartialEq, Arbitrary, Deserialize, Serialize)]
struct InnerTupleStruct(bool, i32, String);

fuzz_target!(|initial: KitchenSink| {
    // TODO: make this error
    // Only normal real numbers are allowed
    if initial.float32.is_normal() && initial.float64.is_normal() {
        let vdf_text = to_string(&initial).unwrap();
        let reparsed = from_str(&vdf_text).unwrap();

        assert_eq!(initial, reparsed,);
    }
});
