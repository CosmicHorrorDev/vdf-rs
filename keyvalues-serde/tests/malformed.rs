use insta::assert_snapshot;
use keyvalues_serde::{error::Result, from_str};
use serde::Deserialize;

use std::{collections::HashMap, fmt};

mod utils;

use crate::utils::{read_asset_file, BoxedResult, Container};

// Helper macro that generates the boilerplate for snapshotting a deserialization error backed by a
// file
macro_rules! test_snapshot_de {
    ($func_name:ident, $de_ty:ty, $file_name:expr $(,)?) => {
        #[test]
        fn $func_name() -> BoxedResult<()> {
            let vdf_text = read_asset_file($file_name)?;
            snapshot_de_err::<$de_ty>(stringify!($func_name), &vdf_text);

            Ok(())
        }
    };
}

fn snapshot_de_err<'a, T: Deserialize<'a> + fmt::Debug>(snapshot_name: &str, vdf_text: &'a str) {
    let result: Result<T> = from_str(vdf_text);
    let err = result.unwrap_err();
    assert_snapshot!(snapshot_name, err.to_string());
}

test_snapshot_de!(
    str_when_wanting_obj,
    Container<HashMap<String, String>>,
    "string_container.vdf",
);

test_snapshot_de!(
    str_when_wanting_top_level_obj,
    Container<String>,
    "top_level_string.vdf",
);

test_snapshot_de!(obj_when_wanting_str, Container<String>, "obj_container.vdf");

#[test]
fn incorrect_seq_length() -> BoxedResult<()> {
    let name_base = "incorrect_seq_length";
    let vdf_len_one = read_asset_file("string_container.vdf")?;
    let name = format!("{}-one_expecting_two", name_base);
    snapshot_de_err::<Container<(String, String)>>(&name, &vdf_len_one);

    let vdf_len_two = read_asset_file("sequence_string_double.vdf")?;
    let name = format!("{}-two_expecting_one", name_base);
    snapshot_de_err::<Container<(String,)>>(&name, &vdf_len_two);
    let name = format!("{}-two_expecting_three", name_base);
    snapshot_de_err::<Container<(String, String, String)>>(&name, &vdf_len_two);

    Ok(())
}

#[test]
fn parsing_obj_as_sequence() {
    let text = r#""Blah" {}"#;
    assert!(from_str::<Vec<String>>(text).is_err());
}

#[derive(Deserialize, Debug)]
pub struct Pair {
    pub first: String,
    pub second: String,
}

test_snapshot_de!(wants_too_many_members, Pair, "string_container.vdf",);

const INVALID_BOOL_TEXT: &str = r#"
"Container"
{
    "inner" "2"
}
"#;

#[test]
fn invalid_bool() {
    snapshot_de_err::<Container<bool>>("invalid_bool", INVALID_BOOL_TEXT);
}

const ZERO_LEN_CHAR_TEXT: &str = r#"
"Container"
{
    "inner" ""
}
"#;

const TWO_LEN_CHAR_TEXT: &str = r#"
"Container"
{
    "inner" "ab"
}
"#;

#[test]
fn invalid_chars() {
    let name_base = "invalid_chars";
    snapshot_de_err::<Container<char>>(&format!("{}-zero_len", name_base), ZERO_LEN_CHAR_TEXT);
    snapshot_de_err::<Container<char>>(&format!("{}-two_len", name_base), TWO_LEN_CHAR_TEXT);
}
