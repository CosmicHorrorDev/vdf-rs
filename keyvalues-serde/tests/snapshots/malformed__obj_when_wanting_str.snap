---
source: keyvalues-serde/tests/malformed.rs
assertion_line: 28
expression: err.to_string()

---
EOF while parsing key or value
